File: Core/Source/NSString+CSS.m
//
//  NSString+CSS.m
//  DTCoreText
//
//  Created by Oliver Drobnik on 31.01.12.
//  Copyright (c) 2012 Drobnik.com. All rights reserved.
//

#import "DTCoreText.h"
#import "NSString+CSS.h"

@implementation NSString (CSS)

#pragma mark CSS

- (NSDictionary *)dictionaryOfCSSStyles
{
	// font-size:14px;
	NSScanner *scanner = [NSScanner scannerWithString:self];
	
	NSString *name = nil;
	NSString *value = nil;
	
	NSMutableDictionary *tmpDict = [NSMutableDictionary dictionary];
	
	@autoreleasepool
	{
		while ([scanner scanCSSAttribute:&name value:&value])
		{
			[tmpDict setObject:value forKey:name];
		}
	}
	
	// converting to non-mutable costs 37.5% of method
	//	return [NSDictionary dictionaryWithDictionary:tmpDict];
	return tmpDict;
}

- (BOOL)isCSSLengthValue
{
	NSScanner *scanner = [NSScanner scannerWithString:self];
	
	NSString *numberStr;
	
	if (![scanner scanCharactersFromSet:[NSCharacterSet cssLengthValueCharacterSet] intoString:&numberStr])
	{
		return NO;
	}
	
	NSString *numberUnitStr;
	if (![scanner scanCharactersFromSet:[NSCharacterSet cssLengthUnitCharacterSet] intoString:&numberUnitStr])
	{
		return YES;
	}
	
	if ([numberUnitStr isEqualToString:@"em"] | [numberUnitStr isEqualToString:@"px"] | [numberUnitStr isEqualToString:@"pt"])
	{
		return YES;
	}
	
	return NO;
}

- (CGFloat)pixelSizeOfCSSMeasureRelativeToCurrentTextSize:(CGFloat)textSize textScale:(CGFloat)textScale
{
	NSUInteger stringLength = [self length];
	unichar *_characters = calloc(stringLength, sizeof(unichar));
	[self getCharacters:_characters range:NSMakeRange(0, stringLength)];
	
	CGFloat value = 0;
	
	BOOL commaSeen = NO;
	BOOL negative = NO;
	NSUInteger digitsPastComma = 0;
	
	NSUInteger i=0;
	
	for (; i<stringLength; i++)
	{
		unichar ch = _characters[i];
		
		if (ch>='0' && ch<='9')
		{
			float digit = (float)(ch-'0');
			value *= 10.0f;
			value += digit;
			
			if (commaSeen)
			{
				digitsPastComma++;
			}
		}
		else if (ch=='.')
		{
			commaSeen = YES;
		}
		else if (ch=='-')
		{
			negative = YES;
		}
		else
		{
			// non-numeric character
			break;
		}
	}
	
	if (commaSeen)
	{
		value /= powf(10.0f, digitsPastComma);
	}
	
	// skip whitespace
	while (i<stringLength && IS_WHITESPACE(_characters[i]))
	{
		i++;
	}
	
	if (i<stringLength)
	{
		unichar ch = _characters[i++];
		
		if (ch == '%')
		{
			// percent value
			value *= textSize / 100.0f;
		}
		else if (ch == 'e')
		{
			if (i<stringLength)
			{
				if (_characters[i] == 'm')
				{
					// em value
					value *= textSize;
				}
			}
		}
		else if (ch == 'p')
		{
			if (i<stringLength)
			{
				if (_characters[i] == 'x')
				{
					// absolute pixel value gets scaled
					value *= textScale;
				}
				else if (_characters[i] == 't')
				{
					// 1 pt = 1.3333 px on Mac, so we do the same
					value *= 1.3333f;
					
					// absolute pixel value gets scaled
					value *= textScale;
				}
			}
		}
	}
	
	if (negative)
	{
		value *= -1;
	}
	
	free(_characters);
	return value;
}

#pragma mark - Margins / Padding

- (DTEdgeInsets)DTEdgeInsetsRelativeToCurrentTextSize:(CGFloat)textSize textScale:(CGFloat)textScale
{
	DTEdgeInsets edgeInsets = {0,0,0,0};
	
	if ([self length])
	{
		// maybe it's using the short style
		NSArray *parts = [self componentsSeparatedByString:@" "];
		
		if ([parts count] == 4)
		{
			edgeInsets.top = [[parts objectAtIndex:0] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets.right = [[parts objectAtIndex:1] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets.bottom = [[parts objectAtIndex:2] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets.left = [[parts objectAtIndex:3] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
		}
		else if ([parts count] == 3)
		{
			edgeInsets.top = [[parts objectAtIndex:0] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets.right = [[parts objectAtIndex:1] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets.bottom = [[parts objectAtIndex:2] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets.left = edgeInsets.right;
		}
		else if ([parts count] == 2)
		{
			edgeInsets.top = [[parts objectAtIndex:0] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets.right = [[parts objectAtIndex:1] pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets.bottom = edgeInsets.top;
			edgeInsets.left = edgeInsets.right;
		}
		else
		{
			CGFloat paddingAmount = [self pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:textScale];
			edgeInsets = DTEdgeInsetsMake(paddingAmount, paddingAmount, paddingAmount, paddingAmount);
		}
	}
	
	return edgeInsets;
}

#pragma mark - CSS Shadows

- (NSArray *)arrayOfCSSShadowsWithCurrentTextSize:(CGFloat)textSize currentColor:(DTColor *)color
{
	NSString *trimmedString = [self stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
	
	if ([trimmedString isEqualToString:@"none"])
	{
		return nil;
	}
	
	NSScanner *scanner = [NSScanner scannerWithString:trimmedString];
	
	NSMutableCharacterSet *tokenEndSet = [[NSCharacterSet whitespaceAndNewlineCharacterSet] mutableCopy];
	[tokenEndSet addCharactersInString:@","];
	
	
	NSMutableArray *tmpArray = [NSMutableArray array];
	
	while (![scanner isAtEnd])
	{
		DTColor *shadowColor = nil;
		
		NSString *offsetXString = nil;
		NSString *offsetYString = nil;
		NSString *blurString = nil;
		
		if ([scanner scanHTMLColor:&shadowColor])
		{
			// format: <color> <length> <length> <length>?
			
			if ([scanner scanUpToCharactersFromSet:tokenEndSet intoString:&offsetXString])
			{
				if ([scanner scanUpToCharactersFromSet:tokenEndSet intoString:&offsetYString])
				{
					// blur is optional
					[scanner scanUpToCharactersFromSet:tokenEndSet intoString:&blurString];
					
					
					CGFloat offset_x = [offsetXString pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:1.0f];
					CGFloat offset_y = [offsetYString pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:1.0f];
					CGSize offset = CGSizeMake(offset_x, offset_y);
					CGFloat blur = [blurString pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:1.0f];
					
					NSValue *offsetValue;
#if TARGET_OS_IPHONE
					offsetValue = [NSValue valueWithCGSize:offset];
#else
					offsetValue = [NSValue valueWithSize:offset];
#endif
					
					NSDictionary *shadowDict = [NSDictionary dictionaryWithObjectsAndKeys:offsetValue, @"Offset",
														 DTNSNumberFromCGFloat(blur), @"Blur",
														 shadowColor, @"Color", nil];
					
					[tmpArray addObject:shadowDict];
				}
			}
		}
		else
		{
			// format: <length> <length> <length>? <color>?
			
			if ([scanner scanUpToCharactersFromSet:tokenEndSet intoString:&offsetXString])
			{
				if ([scanner scanUpToCharactersFromSet:tokenEndSet intoString:&offsetYString])
				{
					// blur is optional
					if (![scanner scanHTMLColor:&shadowColor])
					{
						if ([scanner scanUpToCharactersFromSet:tokenEndSet intoString:&blurString])
						{
							if (![scanner scanHTMLColor:&shadowColor])
							{
								// invalid color, we ignore this color
							}
						}
					}
					
					if (!shadowColor)
					{
						// color is same as color attribute of style
						shadowColor = color;
					}
					
					CGFloat offset_x = [offsetXString pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:1.0f];
					CGFloat offset_y = [offsetYString pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:1.0f];
					CGSize offset = CGSizeMake(offset_x, offset_y);
					CGFloat blur = [blurString pixelSizeOfCSSMeasureRelativeToCurrentTextSize:textSize textScale:1.0f];
					
					NSValue *offsetValue;
#if TARGET_OS_IPHONE
					offsetValue = [NSValue valueWithCGSize:offset];
#else
					offsetValue = [NSValue valueWithSize:offset];
#endif
					
					NSDictionary *shadowDict = [NSDictionary dictionaryWithObjectsAndKeys:offsetValue, @"Offset",
														 DTNSNumberFromCGFloat(blur), @"Blur",
														 shadowColor, @"Color", nil];
					
					[tmpArray addObject:shadowDict];
				}
			}
		}
		
		// now there should be a comma
		if (![scanner scanString:@"," intoString:NULL])
		{
			break;
		}
	}
	
	// only return array if not empty
	if ([tmpArray count])
	{
		return tmpArray;
	}
	
	return nil;
}

- (NSString *)stringByDecodingCSSContentAttribute
{
	NSUInteger length = [self length];
	
	unichar *characters = calloc(length, sizeof(unichar));
	unichar *final = calloc(length, sizeof(unichar));
	
	[self getCharacters:characters range:NSMakeRange(0, length)];
	
	NSUInteger outChars = 0;
	
	BOOL inEscapedSequence = NO;
	unichar decodedChar = 0;
	NSUInteger escapedCharacterCount = 0;
	
	for (NSUInteger idx=0; idx<length;idx++)
	{
		unichar character = characters[idx];
		
		if (inEscapedSequence && (escapedCharacterCount<4))
		{
			if (character=='\\')
			{
				// escaped backslash
				final[outChars++] = '\\';
				inEscapedSequence = NO;
			}
			else if ((character>='0' && character<='9') || (character>='A' && character<='F') || (character>='a' && character<='f'))
			{
				// hex digit
				decodedChar *= 16;
				
				if (character>='0' && character<='9')
				{
					decodedChar += (character - '0');
				}
				else if (character>='A' && character<='F')
				{
					decodedChar += (character - 'A' + 10);
				}
				else if (character>='a' && character<='f')
				{
					decodedChar += (character - 'a' + 10);
				}
				
				escapedCharacterCount++;
			}
			else
			{
				// illegal character following slash
				final[outChars++] = '\\';
				final[outChars++] = character;
				
				inEscapedSequence = NO;
			}
		}
		else
		{
			if (inEscapedSequence)
			{
				// output what we have decoded so far
				final[outChars++] = decodedChar;
			}
			
			if (character == '\\')
			{
				// begin of escape sequence
				decodedChar = 0;
				escapedCharacterCount = 0;
				inEscapedSequence = YES;
			}
			else
			{
				inEscapedSequence = NO;
				
				// just copy
				final[outChars++] = character;
			}
		}
	}
	
	// if string ended in escaped sequence we still need to output
	if (inEscapedSequence)
	{
		// output what we have decoded so far
		final[outChars++] = decodedChar;
	}
	
	free(characters);
	NSString *clean = [[NSString alloc] initWithCharacters:final length:outChars];
	free(final);
	
	return clean;
}

@end

File: Sources/DTCoreTextSwift/NSString+CSS.swift
//
//  NSString+CSS.swift
//  DTCoreText
//
//  Created by Oliver Drobnik on 31.01.12.
//  Copyright (c) 2012 Drobnik.com. All rights reserved.
//

import Foundation

// MARK: - Whitespace Check Helper (local)

private func isWhitespace(_ c: unichar) -> Bool {
    return c == 0x20 || c == 0x09 || c == 0x0A || c == 0x0B || c == 0x0C || c == 0x0D || c == 0x85
}

/// Methods to make dealing with CSS strings easier.
extension NSString {

    /// Examine a string for all CSS styles that are applied to it and return a dictionary of those styles.
    @objc public func dictionaryOfCSSStyles() -> [String: Any] {
        let scanner = Scanner(string: self as String)

        var tmpDict = [String: Any]()

        autoreleasepool {
            var namePtr: NSString?
            var valuePtr: AnyObject?

            while scanner.scanCSSAttribute(&namePtr, value: &valuePtr) {
                if let name = namePtr as String?, let value = valuePtr {
                    tmpDict[name] = value
                }
            }
        }

        return tmpDict
    }

    /// Determines if the receiver contains a CSS length value.
    @objc public func isCSSLengthValue() -> Bool {
        let scanner = Scanner(string: self as String)

        if scanner.scanCharacters(from: NSCharacterSet.dt_cssLengthValueCharacterSet) == nil {
            return false
        }

        guard let unit = scanner.scanCharacters(from: NSCharacterSet.dt_cssLengthUnitCharacterSet) else {
            return true
        }

        if unit == "em" || unit == "px" || unit == "pt" {
            return true
        }

        return false
    }

    /// Calculates a pixel-based length from the receiver based on the current text size in pixels.
    @objc public func pixelSizeOfCSSMeasure(relativeToCurrentTextSize textSize: CGFloat, textScale: CGFloat) -> CGFloat {
        let stringLength = self.length
        guard stringLength > 0 else { return 0 }

        var characters = [unichar](repeating: 0, count: stringLength)
        self.getCharacters(&characters, range: NSRange(location: 0, length: stringLength))

        var value: CGFloat = 0
        var commaSeen = false
        var negative = false
        var digitsPastComma: UInt = 0
        var i = 0

        while i < stringLength {
            let ch = characters[i]

            if ch >= 0x30 && ch <= 0x39 { // '0'-'9'
                let digit = CGFloat(ch - 0x30)
                value *= 10.0
                value += digit

                if commaSeen {
                    digitsPastComma += 1
                }
            } else if ch == 0x2E { // '.'
                commaSeen = true
            } else if ch == 0x2D { // '-'
                negative = true
            } else {
                break
            }
            i += 1
        }

        if commaSeen {
            value /= pow(10.0, CGFloat(digitsPastComma))
        }

        // skip whitespace
        while i < stringLength && isWhitespace(characters[i]) {
            i += 1
        }

        if i < stringLength {
            let ch = characters[i]
            i += 1

            if ch == 0x25 { // '%'
                value *= textSize / 100.0
            } else if ch == 0x65 { // 'e'
                if i < stringLength && characters[i] == 0x6D { // 'm'
                    value *= textSize
                }
            } else if ch == 0x70 { // 'p'
                if i < stringLength {
                    if characters[i] == 0x78 { // 'x'
                        value *= textScale
                    } else if characters[i] == 0x74 { // 't'
                        value *= 1.3333
                        value *= textScale
                    }
                }
            }
        }

        if negative {
            value *= -1
        }

        return value
    }

    /// Decodes edge inset values from the CSS attribute string.
    @objc public func dtEdgeInsets(relativeToCurrentTextSize textSize: CGFloat, textScale: CGFloat) -> DTEdgeInsets {
        var edgeInsets = DTEdgeInsets(top: 0, left: 0, bottom: 0, right: 0)

        if self.length > 0 {
            let parts = self.components(separatedBy: " ")

            if parts.count == 4 {
                edgeInsets.top = (parts[0] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets.right = (parts[1] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets.bottom = (parts[2] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets.left = (parts[3] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
            } else if parts.count == 3 {
                edgeInsets.top = (parts[0] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets.right = (parts[1] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets.bottom = (parts[2] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets.left = edgeInsets.right
            } else if parts.count == 2 {
                edgeInsets.top = (parts[0] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets.right = (parts[1] as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets.bottom = edgeInsets.top
                edgeInsets.left = edgeInsets.right
            } else {
                let paddingAmount = self.pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: textScale)
                edgeInsets = DTEdgeInsets(top: paddingAmount, left: paddingAmount, bottom: paddingAmount, right: paddingAmount)
            }
        }

        return edgeInsets
    }

    /// Parse CSS shadow styles out of this string.
    @objc public func arrayOfCSSShadows(withCurrentTextSize textSize: CGFloat, currentColor color: DTColor?) -> [Any]? {
        let trimmedString = self.trimmingCharacters(in: .whitespacesAndNewlines)

        if trimmedString == "none" {
            return nil
        }

        let scanner = Scanner(string: trimmedString)

        var tokenEndSet = CharacterSet.whitespacesAndNewlines
        tokenEndSet.insert(charactersIn: ",")

        var tmpArray = [[String: Any]]()

        while !scanner.isAtEnd {
            var shadowColor: DTColor?
            var colorPtr: DTColor?

            if scanner.scanHTMLColor(&colorPtr) {
                shadowColor = colorPtr

                // format: <color> <length> <length> <length>?
                if let offsetXString = scanner.scanUpToCharacters(from: tokenEndSet),
                   let offsetYString = scanner.scanUpToCharacters(from: tokenEndSet) {
                    // blur is optional
                    let blurString = scanner.scanUpToCharacters(from: tokenEndSet)

                    let offsetX = (offsetXString as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: 1.0)
                    let offsetY = (offsetYString as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: 1.0)
                    let offset = CGSize(width: offsetX, height: offsetY)
                    let blur = blurString != nil ? (blurString! as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: 1.0) : 0

                    #if os(iOS)
                    let offsetValue = NSValue(cgSize: offset)
                    #else
                    let offsetValue = NSValue(size: offset)
                    #endif

                    var shadowDict: [String: Any] = [
                        "Offset": offsetValue,
                        "Blur": NSNumber(value: Double(blur)),
                    ]
                    if let sc = shadowColor {
                        shadowDict["Color"] = sc
                    }
                    tmpArray.append(shadowDict)
                }
            } else {
                // format: <length> <length> <length>? <color>?
                if let offsetXString = scanner.scanUpToCharacters(from: tokenEndSet),
                   let offsetYString = scanner.scanUpToCharacters(from: tokenEndSet) {
                    // blur is optional
                    var blurString: String?
                    if !scanner.scanHTMLColor(&colorPtr) {
                        blurString = scanner.scanUpToCharacters(from: tokenEndSet)
                        if blurString != nil {
                            _ = scanner.scanHTMLColor(&colorPtr)
                        }
                    }
                    shadowColor = colorPtr

                    if shadowColor == nil {
                        shadowColor = color
                    }

                    let offsetX = (offsetXString as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: 1.0)
                    let offsetY = (offsetYString as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: 1.0)
                    let offset = CGSize(width: offsetX, height: offsetY)
                    let blur = blurString != nil ? (blurString! as NSString).pixelSizeOfCSSMeasure(relativeToCurrentTextSize: textSize, textScale: 1.0) : 0

                    #if os(iOS)
                    let offsetValue = NSValue(cgSize: offset)
                    #else
                    let offsetValue = NSValue(size: offset)
                    #endif

                    var shadowDict: [String: Any] = [
                        "Offset": offsetValue,
                        "Blur": NSNumber(value: Double(blur)),
                    ]
                    if let sc = shadowColor {
                        shadowDict["Color"] = sc
                    }
                    tmpArray.append(shadowDict)
                }
            }

            // now there should be a comma
            if scanner.scanString(",") == nil {
                break
            }
        }

        return tmpArray.isEmpty ? nil : tmpArray
    }

    /// Decodes a content attribute which might contain unicode sequences.
    @objc public func stringByDecodingCSSContentAttribute() -> String {
        let length = self.length
        guard length > 0 else { return "" }

        var characters = [unichar](repeating: 0, count: length)
        self.getCharacters(&characters, range: NSRange(location: 0, length: length))

        var final_ = [unichar](repeating: 0, count: length)
        var outChars = 0

        var inEscapedSequence = false
        var decodedChar: unichar = 0
        var escapedCharacterCount: UInt = 0

        for idx in 0..<length {
            let character = characters[idx]

            if inEscapedSequence && escapedCharacterCount < 4 {
                if character == 0x5C { // '\\'
                    final_[outChars] = 0x5C
                    outChars += 1
                    inEscapedSequence = false
                } else if (character >= 0x30 && character <= 0x39) || // 0-9
                          (character >= 0x41 && character <= 0x46) || // A-F
                          (character >= 0x61 && character <= 0x66) {  // a-f
                    decodedChar &*= 16

                    if character >= 0x30 && character <= 0x39 {
                        decodedChar &+= character &- 0x30
                    } else if character >= 0x41 && character <= 0x46 {
                        decodedChar &+= character &- 0x41 &+ 10
                    } else if character >= 0x61 && character <= 0x66 {
                        decodedChar &+= character &- 0x61 &+ 10
                    }

                    escapedCharacterCount += 1
                } else {
                    // illegal character following slash
                    final_[outChars] = 0x5C
                    outChars += 1
                    final_[outChars] = character
                    outChars += 1
                    inEscapedSequence = false
                }
            } else {
                if inEscapedSequence {
                    // output what we have decoded so far
                    final_[outChars] = decodedChar
                    outChars += 1
                }

                if character == 0x5C { // '\\'
                    decodedChar = 0
                    escapedCharacterCount = 0
                    inEscapedSequence = true
                } else {
                    inEscapedSequence = false
                    final_[outChars] = character
                    outChars += 1
                }
            }
        }

        // if string ended in escaped sequence we still need to output
        if inEscapedSequence {
            final_[outChars] = decodedChar
            outChars += 1
        }

        return String(utf16CodeUnits: final_, count: outChars)
    }
}
